Software/src/LedDeviceWarls.cpp

```diff
@@ -0,0 +1,141 @@
+/*
+ * LedDeviceWarls.cpp
+ *
+ *	Created on: 06.09.2011
+ *		Author: Mike Shatohin (brunql), Timur Sattarov
+ *		Project: Lightpack
+ *
+ *	Lightpack is very simple implementation of the backlight for a laptop
+ *
+ *	Copyright (c) 2011 Mike Shatohin, mikeshatohin [at] gmail.com
+ *
+ *	Lightpack is free software: you can redistribute it and/or modify
+ *	it under the terms of the GNU General Public License as published by
+ *	the Free Software Foundation, either version 2 of the License, or
+ *	(at your option) any later version.
+ *
+ *	Lightpack is distributed in the hope that it will be useful,
+ *	but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the
+ *	GNU General Public License for more details.
+ *
+ *	You should have received a copy of the GNU General Public License
+ *	along with this program.	If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "LedDeviceWarls.hpp"
+#include "enums.hpp"
+
+LedDeviceWarls::LedDeviceWarls(const QString& address, const QString& port, const int timeout, QObject * parent) : AbstractLedDeviceUdp(address, port, timeout, parent)
+{
+}
+
+const QString LedDeviceWarls::name() const
+{ 
+	return "warls"; 
+}
+
+int LedDeviceWarls::maxLedsCount()
+{
+	return MaximumNumberOfLeds::Warls;
+}
+
+void LedDeviceWarls::setColors(const QList<QRgb> & colors)
+{
+	if(colors.size()> 0)
+	{
+		resizeColorsBuffer(colors.count());
+
+		applyColorModifications(colors, m_colorsBuffer);
+
+		m_writeBuffer.clear();
+		m_writeBuffer.append(m_writeBufferHeader);
+
+		for (int i = 0; i < m_colorsBuffer.count(); i++)
+		{
+			if (colors[i] != m_colorsSaved[i])
+			{
+				StructRgb color = m_colorsBuffer[i];
+
+				// Reduce 12-bit colour information
+				color.r = color.r >> 4;
+				color.g = color.g >> 4;
+				color.b = color.b >> 4;
+
+				m_writeBuffer.append(i);
+				m_writeBuffer.append(color.r);
+				m_writeBuffer.append(color.g);
+				m_writeBuffer.append(color.b);
+			}
+		}
+
+		m_colorsSaved = colors;
+	}
+
+	// This may send the header only
+	bool ok = writeBuffer(m_writeBuffer);
```

Comment:
the `write` should be inside the `if ..` otherwise the buffer isn't getting reset

Reply:
That was something I seem to have picked up from the Virtual Device, I notice Ada and Ardu don't have that pattern, should it just be removed instead?